### PR TITLE
grpclb: switch to fallback mode if all connections are lost

### PR DESCRIPTION
**Pull request by zhangkun83**
Previously fallback mode can be entered only if the client has not
received any server list and the fallback timeout has expired.

Now the fallback timer is started when the stream to the balancer is broken
AND there is no ready Subchannels. Fallback mode is activated when the
timer expires. When a new server list is received from the balancer, either
the fallback timer is cancelled, or fallback mode is exited.

Also fixed a bug that the fallback timer should've been cancelled when GrpcState
is shut down.

### CODE REVIEW

**Comment by zhangkun83** [comment]
@markdroth please focus on the behavior.
@dapengzhang0 please focus on the language.

File: grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java [hunk 1]
```diff
@@ -1311,27 +1317,132 @@ public void grpclbFallbackToBackendsFromResolver() {
...
+    List<ServerEntry> serverList = Arrays.asList(
+        new ServerEntry("127.0.0.1", 2000, "token0001"),
+        new ServerEntry("127.0.0.1", 2010, "token0002"));
+    lbResponseObserver.onNext(buildInitialResponse());
+    lbResponseObserver.onNext(buildLbResponse(serverList));
+
+    List<Subchannel> subchannels =
+        fallbackTestVerifyUseOfBalancerBackendLists(inOrder, helper, serverList);
+
+    // Let fallback timer expire
```
Comment (markdroth) [inline grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java]: I don't know this code well enough to understand exactly what this is covering, so let me say what the behavior should be here and let you tell me if that's what the code is doing.

I think the behavior we want here is for the fallback timer to start again at the moment that we lose contact with the balancers and backends.  If we don't reestablish a connection with either a balancer or a backend by the time the timer expires, then we use the fallback addresses until we regain contact with the balancer.

Note that the above is different from the initial fallback timer that we had implemented previously.  For example, consider the following scenario:

1. Channel is created and is moved out of idle state (either by starting an RPC or by requesting that the channel start connecting).
2. We start trying to connect to the balancer.  At the same time, we start the fallback timer.
3. If the fallback timer expires before we connect to the balancer, we use the fallback addresses while we continue to attempt to connect to the balancers.
4. We successfully connect to the balancer and receive a serverlist.  If the fallback timer had not previously expired, we cancel it; if it had previously expired, we stop using the fallback addresses.  We then start using the backends that the balancer told us about.
5. Some time later, we lose contact with the balancer.  However, we are still in contact with the backends from the last serverlist the balancer sent us, so we keep using them while we try to reconnect to the balancer.
6. Now we lose contact with all backends from the last serverlist the balancer sent us.  We stat a new fallback timer.
7. If we regain contact with the balancer and get a new serverlist before the fallback timer expires, then we cancel the timer and use the new serverlist.
8. If the timer fires before we regain contact with the balancer, then we switch to using the fallback addresses until we regain contact with the balancer.

In other words, there are two different scenarios in which the fallback timer will be used.  The one that we had previously implemented is a timer that starts when we start trying to contact the balancer.  The new one that we need in this PR will start at the moment when we lose contact with the last backend if we have also lost contact with the balancer.

Comment (zhangkun83) [reply grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java]: In the current code the timeout is used only for the initial fallback. Later fallbacks are not subject to the timeout, thus will activate immediately after all connections are lost (as long as the initial timeout is passed).

We should discuss and decide which approach we want to take.

My understanding is that the initial timeout is necessary because the client needs time to connect to balancer and receive the server list. It doesn't seem to be the case for the later fallbacks.  Losing connections to the balancer and all backends is pretty unusual, and I don't see a reason why we still have to wait for 10 seconds instead of using the fallbacks immediately.

Comment (zhangkun83) [reply grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java]: @markdroth I have updated the PR to add the timeout to subsequent fallbacks. Now it matches your description.

Comment (markdroth) [reply grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java]: Sounds great!  Then feel free to merge this PR as soon as Penn reviews it.  Thanks!

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java [hunk 2]
```diff
@@ -142,7 +142,7 @@ public void handleResolvedAddressGroups(
           grpclbState.propagateError(Status.UNAVAILABLE.withDescription(
                   "NameResolver returned no LB address while asking for GRPCLB"));
         } else {
-          grpclbState.updateAddresses(newLbAddressGroups, newBackendServers);
+          grpclbState.handleAddresses(newLbAddressGroups, newBackendServers);
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java]: Done.

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java [hunk 3]
```diff
@@ -158,30 +162,56 @@ void updateAddresses(
...
-  private void maybeUseFallbackBackends() {
-    // Only use fallback backends after fallback timer expired and before receiving server list from
-    // the balancer.
-    if (receivedServerListFromBalancer || !fallbackTimerExpired) {
-      return;
+  /**
+   * Start the fallback timer if it's not already started and all connections are lost.
+   */
+  private void maybeStartFallbackTimer() {
+    if (fallbackTimer == null) {
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java]: Done.

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java [hunk 4]
```diff
@@ -158,30 +162,56 @@ void updateAddresses(
...
+      if (fallbackBackendList.isEmpty()) {
+        return;
+      }
+      if (balancerWorking) {
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java]: Nice catch! Fixed and added test for it.

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java [hunk 5]
```diff
@@ -158,30 +162,56 @@ void updateAddresses(
...
+        if (subchannel.getAttributes().get(STATE_INFO).get().getState() == READY) {
+          numReadySubchannels++;
+        }
+      }
+      if (numReadySubchannels > 0) {
+        return;
+      }
+      logger.log(Level.FINE, "[{0}] Starting fallback timer.", new Object[] {logId});
+      fallbackTimer = new FallbackModeTask();
+      fallbackTimer.scheduledFuture =
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java]: Done.
